### ISSUE TITLE
[feature] Having option to not host the TD

### ISSUE THREAD

**Issue by egekorkan** [eclipse/thingweb.node-wot#195]
Currently, the generated TDs by an Exposed Thing are always hosted by the Servient. It is a bit of a security concern if the TD is guaranteed to be on the device. I would say that there should be an option settable somewhere (in the servient level?) that disables the hosting of the TD. This would have to be implemented for all the bindings though.

**Comment by relu91** [comment]
It is a nice feature! but I do not completely understand the security concerns you have about hosting the TD on the device. Could you please extend a little bit to that point? Do you imply that the TD could be stored inside a super-secure facility, while the device might have a less constricted security requirements?

**Comment by egekorkan** [comment]
From one point of view, TD removes "security by obscurity" which is not a good security choice anyways but somewhat justifiable. It should be possible that the developer of the Thing wants to hide the TD for every user who wants to communicate with the device or maybe provide only a part of the TD to some users. 

It could be also seen from a usability perspective where a TD has a lot of interactions and only some of them are relevant for most users but the rest can be relevant for more "advanced users". This would need to be managed by a TD directory which is doable at the moment too.

From another point of view, in node-wot security is applied only to the interactions but there is actually no id/password needed to fetch a TD that has basic auth. When you apply the following configuration to counter.js example, the TD is still reachable without any id/password:
```json
{
    "http": {
        "port": 8080,
        "serverKey": "../security/privatekey.pem",
        "serverCert": "../security/certificate.pem",
        "security": {
            "scheme": "basic"
        }
    }
}
```
This might be intended by design since the security is for interactions but TD is served by the servient. Furthermore, this makes sense when you think that security configuration is defined for each protocol but the TD contains all the protocols.

Going a bit further:

If the security configurations were to be applied to fetching a TD, if I set the HTTP to `nosec`  and CoAP to `apikey`: 

1. Should the TD be reachable with no security over HTTP but have also the CoAP forms? -> Security "breach" for CoAP
2. Should the TD be reachable with no security over HTTP but not have the CoAP forms? -> Makes sense, but this is a mechanism that needs quite a bit of implementation since every Thing has one TD for node-wot. There, the CoAP TD can be linked to by using the links array but this is also not implemented. See issue #160.


By the way, this was pointed out by @jenlincode but I might have taken it a bit further :)

**Comment by danielpeintner** [comment]
Note: having an option to report *no TD*  for node-wot would not be a big deal I believe.

However, I wonder which consequence it would have on the scripting *IF* the TD is no longer reported by the thing itself. At the moment `ConsumedThing`/`ExposedThing` has the method call [`getThingDescription()`](https://w3c.github.io/wot-scripting-api/#dom-consumedthing-getthingdescription).

At the moment the TD can be hosted elsewhere also but it was kind of assumed it is hosted by the TD itself as well...

**Comment by relu91** [comment]
Thanks @egekorkan it is much more clear now and I agree with your considerations. 

>It could be also seen from a usability perspective where a TD has a lot of interactions and only some of them are relevant for most users but the rest can be relevant for more "advanced users". This would need to be managed by a TD directory which is doable at the moment too.

Just to make an example in a real use-case. In our lab, we are using two different TDs for each sensor. One contains only "read" operations and it is designed with a low-security level, while the other is used for sensor configuration and it is protected with higher level of security. Ask @filobotto for more details.

>However, I wonder which consequence it would have on the scripting IF the TD is no longer reported by the thing itself. At the moment ConsumedThing/ExposedThing has the method call `getThingDescription()`.

I guess it wouldn't be a problem for ConsumedThing/ExposedThing. If you are able to Expose or Consume a Thing with the "hosted-somewhere else" TD it means that the script has the security rights to read it. or I might miss something in your statement.

**Comment by egekorkan** [comment]
I also agree with @relu91, all this would be after a successful fetch of the TD and does not have any influence on the Scripting API
